src/uct/tcp/tcp_ep.c

```diff
@@ -834,20 +834,75 @@ static inline void uct_tcp_ep_am_send(uct_tcp_iface_t *iface, uct_tcp_ep_t *ep,
 
     uct_tcp_ep_send(ep);
 
-    if (uct_tcp_ep_ctx_buf_need_progress(&ep->tx)) {
-        uct_tcp_ep_mod_events(ep, UCS_EVENT_SET_EVWRITE, 0);
-    } else {
+    if (!uct_tcp_ep_ctx_buf_need_progress(&ep->tx)) {
         uct_tcp_ep_ctx_reset(&ep->tx);
+    } else {
+        uct_tcp_ep_mod_events(ep, UCS_EVENT_SET_EVWRITE, 0);
     }
 }
 
+static inline void
+uct_tcp_ep_am_short_fill_data(uct_tcp_am_hdr_t *hdr, uint64_t header,
+                              const void *payload, unsigned length)
+{
+    *((uint64_t*)(hdr + 1)) = header;
+    memcpy(UCS_PTR_BYTE_OFFSET(hdr + 1, sizeof(header)), payload, length);
+}
+
+static const void*
+uct_tcp_ep_am_sendv_trace_payload(uct_tcp_am_hdr_t *hdr,
+                                  const void *header,
+                                  const struct iovec *payload_iov,
+                                  int short_sendv)
+{
+    if (!short_sendv) {
+        return header;
+    }
+
+    /* If user requested trace data, we copy header and payload
+     * to EP TX buffer in order to trace correct data */
+    uct_tcp_ep_am_short_fill_data(hdr, *((const uint64_t*)header),
+                                  payload_iov->iov_base, payload_iov->iov_len);
+    return (hdr + 1);
+}
+
+static inline ucs_status_t
+uct_tcp_ep_am_sendv(uct_tcp_iface_t *iface, uct_tcp_ep_t *ep,
+                    int short_sendv, uct_tcp_am_hdr_t *hdr,
+                    size_t send_limit, const void *header,
+                    struct iovec *iov, size_t iov_cnt)
+{
+    ucs_status_t status;
+
+    uct_iface_trace_am(&iface->super, UCT_AM_TRACE_TYPE_SEND, hdr->am_id,
+                       /* the function will be invoked only in case of
+                        * data tracing is enabled */
+                       uct_tcp_ep_am_sendv_trace_payload(hdr, header,
+                                                         &iov[2], short_sendv),
+                       hdr->length, "SEND %p fd %d", ep, ep->fd);
+
+    ep->tx.length = hdr->length + sizeof(*hdr);
+
+    ucs_assertv(ep->tx.length <= send_limit, "ep=%p", ep);
+
+    status = ucs_socket_sendv_nb(ep->fd, iov, iov_cnt,
+                                 &ep->tx.offset, NULL, NULL);
+
+    ucs_trace_data("tcp_ep %p: sendv %zu bytes", ep, ep->tx.offset);
```

Comment:
do we really need this trace, in addition to uct_iface_trace_am?

Reply:
i think we should unite this with the other trace message, so there will be one trace per one send

Reply:
done for all places